chore: fix RUSTSEC-2024-0370 proc-macro-error is unmaintained

oleonardolima opened the pull request:
A candidate for fixing bitcoindevkit/bdk#1593. It builds on top of bitcoindevkit/bdk#1615 
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->

In https://github.com/bitcoindevkit/bdk/issues/1593 it's mentioned that `proc-macro-error` is unmaintained for the
past few years, with no fix other than using proc-macro-error2 instead.

As on our scenario it's merely a transitive dependency of `clap`,
through `clap_derive` feature, which in latest releases doesn't depend on
`proc-macro-error` we can just bump it to latest.

It's valid to note that by bumping it, both examples that relies on clap
are no longer MSRV (1.63) compliant.

That said, this PR does:

- Standardize the example packages to have `example_` prefix.
- Exclude examples from running in main `Build & Test` CI job.
- Add new testing step to `Build & Test Example` CI job.
- Bumps the `clap` to `4.5.17`.

### Notes to the reviewers

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

- Standardize the example packages to have `example_` prefix.
- Exclude examples from running in main `Build & Test` CI job.
- Add new testing step to `Build & Test Example` CI job.
- Bumps the `clap` to `4.5.17`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

<!-- 
#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature
--> 

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

oleonardolima committed:
chore(examples)!: update all examples to have `example_` prefix

oleonardolima committed:
refactor(ci)!: update CI to build and test example in specific job

oleonardolima committed:
fix(RUSTSEC-2024-0370)!: bump `clap` to latest, removing transitive
dependency on `proc-macro-error`.

In #1593 it's mentioned that `proc-macro-error` is unmaintained for the
past few years, with no fix other than using proc-macro-error2 instead.

As on our scenario it's merely a transitive dependency of `clap`,
through `clap_derive` feature, which in latest releases doesn't depend on
`proc-macro-error` we can just bump it to latest.

It's valid to note that by bumping it, both examples that relies on clap
are no longer MSRV (1.63) compliant.

oleonardolima commented:
I also wonder if this could become even simpler by adopting [Examples Cargo Target](https://doc.rust-lang.org/cargo/reference/cargo-targets.html?highlight=examples#examples).

ValuedMammal commented on .github/workflows/cont_integration.yml:
```diff
@@ -50,9 +50,9 @@ jobs:
           cargo update -p tokio-util --precise "0.7.11"
       - name: Build
-        run: cargo build ${{ matrix.features }}
+        run: cargo build --workspace --exclude 'example_*' ${{ matrix.features }}
```
We could try just doing `--all-features` 🤔

notmandatory replied:
I think in the past we ran into errors that only happened when we built with `--no-default-features` so I'd like to keep it.

ValuedMammal replied:
keep which one?

notmandatory replied:
I mean keep testing with `--no-default-features` and `--all-features`. I don't mind only building the examples with the current `stable` rust though if that's the issue.

oleonardolima replied:
I think it's better to keep testing with both.

oleonardolima commented:
I think I also found a bug in how the `bdk_esplora` features were structured, it's somewhat depending on `blocking` feature being set up through the `wallet_esplora` example, and that's why it's failing in the CI when excluding the examples from the workspace.

oleonardolima commented:
> I think I also found a bug in how the `bdk_esplora` features were structured, it's somewhat depending on `blocking` feature being set up through the `wallet_esplora` example, and that's why it's failing in the CI when excluding the examples from the workspace.

Also, it's something that we'd find out through bitcoindevkit/bdk#1944 if miniscript std/no-std weren't a blocker.

ValuedMammal commented on .github/workflows/cont_integration.yml:
```diff
@@ -172,3 +172,6 @@ jobs:
         run: cargo build
+      - name: Test
+        working-directory: example-crates/${{ matrix.example-dir }}
+        run: cargo test
```
Just noting I don't think there are any tests here to run.

oleonardolima replied:
Oh, yes. Thank you!

oleonardolima commented:
> I think I also found a bug in how the `bdk_esplora` features were structured, it's somewhat depending on `blocking` feature being set up through the `wallet_esplora` example, and that's why it's failing in the CI when excluding the examples from the workspace.

I was testing the fixes and got it in 8752555, but I'll move it to a separate PR to make it easier to review (#1615).

oleonardolima committed:
fix(bdk_esplora): build with `--no-default-features`

- add `blocking-https` as one of the default features, instead of
  `blocking-https-rustls`, they are basically the same in
  `esplora-client`.
- add `async` and `blocking as required features for each test, using
  the `[[test]]` cargo target.

oleonardolima committed:
fix(bdk_electrum): build with `--no-default-features`

- add `use-rustls` as required features for `test_electrum.rs`, using
  the `[[test]]` cargo target approach.

ValuedMammal commented:
Rather than rename the example-crates can't we use the "exclude" key in the workspace Cargo.toml to exclude them from invocations of `cargo test --workspace`?

https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields

oleonardolima commented:
> https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields

I thought about it, but AFAICT it'll exclude from the workspace for any `cargo` command, and not only for the `cargo test` one.

ValuedMammal reviewed (approved):
After considering the options I don't see any downside to using this solution.

oleonardolima commented:
> After considering the options I don't see any downside to using this solution.

I rebased this one, as I updated the approach in bitcoindevkit/bdk#1615.

oleonardolima committed:
chore: use path as `name` for cargo test targets

ValuedMammal reviewed (approved):
ACK d802d00a06035bec582eb5aadf0281f26bff6acd
